### PR TITLE
[CLOUDTRUST-2829] Fix logout issue when token has expired

### PR DESCRIPTION
**Pull request by fperot74**
I just added a condition in LogoutEndpoint#142...
Other changes are imported code from recent Keycloak code.

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java [hunk 1]
```diff
@@ -135,17 +136,18 @@ public Response logout(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PARAM) String
...
-            Response response =  AuthenticationManager.browserLogout(session, realm, authResult.getSession(), session.getContext().getUri(), clientConnection, headers, initiatingIdp);
-            logger.debug("finishing OIDC browser logout");
-            return response;
-        } else if (userSession != null) { // non browser logout
-            event.event(EventType.LOGOUT);
-            AuthenticationManager.backchannelLogout(session, realm, userSession, session.getContext().getUri(), clientConnection, headers, true);
-            event.user(userSession.getUser()).session(userSession).success();
+            return initiateBrowserLogout(userSession, redirect, state, initiatingIdp);
+        } else if (userSession != null) {
+            // identity cookie is missing but there's valid id_token_hint which matches session cookie => continue with browser logout
```
Comment (harture) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java:141]: je pense que ce commentaire s'applique a la ligne 143

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java [hunk 2]
```diff
@@ -135,17 +136,18 @@ public Response logout(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PARAM) String
...
+            if (UserSessionModel.State.LOGGED_OUT!=userSession.getState()) {
```
Comment (harture) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java]: On pourrait merger cette condition avec celle de la ligne 140 ?

**Review by harture — APPROVED** [review]

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java (hunk 2 above)
Comment (lagess) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java]: small code formatting glitch (spaces around !=)

File: services/src/main/java/org/keycloak/services/util/CookieHelper.java [hunk 3]
```diff
@@ -38,6 +39,8 @@
  */
 public class CookieHelper {
 
+    public static final String LEGACY_COOKIE = "_LEGACY";
```
Comment (lagess) [inline services/src/main/java/org/keycloak/services/util/CookieHelper.java:42]: Some java doc would be nice, in order to know where this _LEGACY comes from

**Review by lagess — APPROVED** [review]

File: services/src/main/java/org/keycloak/services/util/CookieHelper.java (hunk 3 above)
Comment (fperot74) [reply services/src/main/java/org/keycloak/services/util/CookieHelper.java:42]: Sorry... Just a copy/paste from recent Keycloak code. I tried it but it didn't solve the issue. I kept it as it makes things a little bit better.

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java (hunk 2 above)
Comment (fperot74) [reply services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java:142]: 👍

**Pull request merged**